### ISSUE TITLE
[Question]Does hdfs sink connector support putting data to single path?

### ISSUE THREAD

**Issue by mathsigit** [oharastream/ohara#1068]
Hi folks,

We want to create hive table with the path which putting data by OharaStream hdfs sink connector, and hive only support one single folder in hdfs.

If the amount of partition of kafka topic more than one, the OharaStream hdfs sink connector would write data to the different path in hdfs named with partitions.

Could hdfs sink connector support putting data to single path in the feature?

**Comment by chia7712** [comment]
> single path

folder? or file?

**Comment by chia7712** [comment]
ping @jackyoh

**Comment by jackyoh** [comment]
1. OharaStream HDFS Sink Connector 寫入的資料格式如下：
```
/data/topicName/partition-0/flush00001-00005.txt
/data/topicName/partition-0/flush00006-000010.txt
/data/topicName/partition-1/flush00001-000005.txt
```
2. 我想 @mathsigit 所說的 single path 為 folder

**Comment by chia7712** [comment]
> 所說的 single path 為 folder

 add a flag to hdfs connector?

**Comment by vitojeng** [comment]
為了 performance, sink connector 會依 partition 平行寫入不同 hdfs folder, 因此要寫入同一個資料夾有困難, performance 會一落千丈. 我想比較好的方式可能會是後續做 merge...

**Comment by chia7712** [comment]
> 為了 performance, sink connector 會依 partition 平行寫入不同 hdfs folder, 因此要寫入同一個資料夾有困難, performance 會一落千丈

Different (connector) task still output data to different **files** for different partitions. And the change we make is to redirect all files into same **folder**.

for example:

before
```
/data/topicName/partition-0/flush00001-00005.txt
/data/topicName/partition-0/flush00006-000010.txt
/data/topicName/partition-1/flush00001-000005.txt
```

after
```
/data/topicName/flush00001-00005-partition-0.txt
/data/topicName/flush00006-000010-partition-1.txt
/data/topicName/flush00001-000005-partition-2.txt
```

**Comment by vitojeng** [comment]
That's good idea. Does this change fulfill the Hive requirement?

**Comment by jackyoh** [comment]
這樣做之後 Hive 在建立 external table 就可以查詢到 HDFS 的資料

**Comment by chia7712** [comment]
> 這樣做之後 Hive 在建立 external table 就可以查詢到 HDFS 的資料

要麻煩評估一下技術難度以及要改的程式碼幅度喔

@mathsigit @alan791205 當然也非常歡迎你們貢獻！！！

**Comment by jackyoh** [comment]
Ok, 之後會評估需要修改程式的地方以及會遭遇到的問題

**Comment by mathsigit** [comment]
感謝各位大大的討論, 

總體來說, 只要Sink connector可以讓使用者設定 `預計要flush的HDFS上最終路徑`以下簡稱`flush_path`, 不必一落地hdfs就寫入, 可依時間區間(interval time)或是其它規則另外將檔案搬移至flush_path, 原來會依partition分路徑寫入hdfs的規則保留不變, 這樣可以兼顧sink connector的寫入效能.

**Comment by chia7712** [comment]
> 可依時間區間(interval time)或是其它規則另外將檔案搬移至flush_path,

This produces another data consistence issue. For example, hdfs connector has to handle the failure of moving file.

It seems to me that hdfs connector should offer a strategy to user to **choose** how hdfs connector deploy the output files. For example:
```
strategy: by_partitions (default). output path: $topicName/$partition/$file
```
```
strategy: by_topic (default). output path: $topicName/$file
```

**Comment by mathsigit** [comment]
> It seems to me that hdfs connector should offer a strategy to user to choose how hdfs connector deploy the output files.

It would be more better to user that offer a strategy to deploy the output files, and I hope that OharaStream could provider different strategy to choose(for example [Confluent- hdfs connector](https://docs.confluent.io/current/connect/kafka-connect-hdfs/configuration_options.html#partitioner)) some day. 😃

**Comment by chia7712** [comment]
> OharaStream could provider different strategy to choose(for example Confluent- hdfs connector) some day. smiley

there is a issue (#1048) which offers a abstract csv connector. I feel this improvement should be implemented by the issue (#1048)

**Comment by alan791205** [comment]
I think this should cover all sink connectors instead of only csv's.

**Comment by chia7712** [comment]
> I think this should cover all sink connectors instead of only csv's.

what is **this**? Not all sink conenctor output stuff to **filesystem**. I don't think the strategy we discussed is suitable to all sink connectors.

**Comment by alan791205** [comment]
My bad :( , I am not thinking thoroughly when I commented that,
but please let me rephrase, if the purpose is to create external hive table on some kind of format created by HDFS sink connector, it should cover other formats'(hive recognizable) connectors. Or else you want to keep this scope small that only csv connectors can use this method.

**Comment by chia7712** [comment]
>  Or else you want to keep this scope small that only csv connectors can use this method.

I prefer to keep a **small** issue if the csv format is enough to be loaded to hive.

**Comment by chia7712** [comment]
@eechih Could you confirm whether this issue is addressed?

**Comment by eechih** [comment]
> OharaStream could provider different strategy to choose(for example Confluent-hdfs connector) 

@chia7712 @mathsigit  This issue has not been addressed yet. 未來幾個月，V1.0 新的 UI 是較重要的工作項目，因此，這個 Issue 會在新 UI 完成後再進行規劃安排。 

補充：The abstract CSV sink connector 寫入檔案路徑的方式，目前只有一種，規則如下：

/${outputFolder}/${topicName}/partition${partition}/${topicName}-${partition}-${startOffset}.csv

Examples:
- /output/mytopic/partition0/mytopic-0-000000000.csv
- /output/mytopic/partition0/mytopic-0-000001000.csv
- /output/mytopic/partition1/mytopic-1-000000000.csv
- /output/mytopic/partition1/mytopic-1-000001000.csv

**Comment by vitojeng** [comment]
> /${outputFolder}/${topicName}/partition${partition}/${topicName}-${partition}-${startOffset}.csv

maybe typo ?

> /${outputFolder}/${topicName}/${partition}/${topicName}-${partition}-${startOffset}.csv

**Comment by eechih** [comment]
@vitojeng  It isn't typo. The ${partition} variable is a number.

**Comment by vitojeng** [comment]
Got it
